### PR TITLE
use separate types for bounded iterators

### LINKED ISSUE #8: Implement `ExactSizeIterator`
Is there a reason that `map::{Iter, IterMut}` have two fields, `remaining_min` and `remaining_max`? It appears that these iterators (and `map::{Keys, Values}` and `set::Iter`) can implement `ExactSizeIterator`.

### PR DESCRIPTION
**Pull request by apasel422**
closes #8

### CODE REVIEW

**Comment by FlashCat** [comment]
r? @Gankro

(I've picked a reviewer for you, use r? to override)

**Comment by apasel422** [comment]
The failure was due to an upstream problem in `rand`.

**Comment by apasel422** [comment]
Ping.

File: src/map.rs [hunk 1]
```diff
@@ -1167,10 +1167,7 @@ macro_rules! iterator_impl {
                                                 write_ptr = write_ptr.offset(1);
                                             }
                                             External(key, ref $($mut_)* value) => {
-                                                self.remaining_max -= 1;
-                                                if self.remaining_min > 0 {
-                                                    self.remaining_min -= 1;
-                                                }
+                                                self.remaining -= 1;
```
Comment (Gankra) [inline src/map.rs:1170]: Any concern of underflow here?

Comment (apasel422) [inline src/map.rs:1170]: There shouldn't be, since this iterator now tracks the exact number of elements remaining.

**Pull request merged**